lib/Borel/Types/UOM.hs

```diff
@@ -0,0 +1,225 @@
+-- made available to you by its authors as open source software:
+-- you can redistribute it and/or modify it under the terms of
+-- the 3-clause BSD licence.
+--
+-- /Description/
+-- This module defines unit of measurement for metric reports.
+--
+
+{-# OPTIONS -fno-warn-missing-signatures #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TupleSections     #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+module Borel.Types.UOM
+  ( -- * Unit of measurement
+    UOM(..), BaseUOM(..), Prefix(..)
+
+    -- * Convenient constructors
+  , sec, nanosec, byte, megabyte, gigabyte
+
+    -- * Utilities
+  , pUOM, pPrefixUOM, pBaseUOM
+  , convert
+  , nanosecToSec, byteToGigabyte
+  ) where
+
+import Data.Maybe
+import           Control.Applicative
+import           Control.Error.Util
+import           Control.Lens         (Prism', Traversal', preview, prism', re,
+                                       review, (^.), (^?))
+import           Control.Lens.Plated
+import           Control.Monad
+import           Data.Aeson
+import qualified Data.Attoparsec.Text as AT
+import           Data.Monoid
+import           Data.MultiSet        (MultiSet)
+import qualified Data.MultiSet        as S
+import           Data.Text            (Text)
+import qualified Data.Text            as T
+import           Data.Word
+
+nanosec  = UOM Nano Second
+sec      = UOM Base Second
+byte     = UOM Base Byte
+megabyte = UOM Mega Byte
+gigabyte = UOM Giga Byte
+
+data UOM
+  = UOM Prefix BaseUOM
+  | Times UOM UOM
+  deriving (Eq, Ord)
+
+data Prefix
+  = Base
+  | Giga
+  | Nano
+  | Mebi
+  | Mega
+  deriving (Eq, Ord)
+
+data BaseUOM
+  = Second
+  | Hour
+  | Byte
+  | Instance
+  | IPAddress
+  | CPU
+  | VCPU
+  deriving (Eq, Ord)
+
+data ComparisonBase
+  = CTime
+  | CData
+  | CInstance
+  | CIPAddress
+  | CCPU
+  | CVCPU
+  deriving (Eq, Ord)
+
+pBaseUOM :: Prism' Text BaseUOM
+pBaseUOM = prism' pretty parse
+  where pretty Second    = "s"
+        pretty Hour      = "h"
+        pretty Byte      = "B"
+        pretty Instance  = "instance"
+        pretty IPAddress = "ip-address"
+        pretty CPU       = "cpu"
+        pretty VCPU      = "vcpu-allocation"
+        parse "s"               = Just Second
+        parse "h"               = Just Hour
+        parse "B"               = Just Byte
+        parse "instance"        = Just Instance
+        parse "ip-address"      = Just IPAddress
+        parse "cpu"             = Just CPU
+        parse "vcpu-allocation" = Just VCPU
+        parse _                 = Nothing
+
+pPrefixUOM :: Prism' Text Prefix
+pPrefixUOM = prism' pretty parse
+  where pretty Base = ""
+        pretty Giga = "G"
+        pretty Nano = "n"
+        pretty Mebi = "Mi"
+        pretty Mega = "M"
+        parse ""   = Just Base
+        parse "G"  = Just Giga
+        parse "n"  = Just Nano
+        parse "Mi" = Just Mebi
+        parse "M"  = Just Mega
+        parse _    = Nothing
+
+pUOM :: Prism' Text UOM
+pUOM = prism' pretty parse
```

Comment:
well, here I'm relying on prior-knowledge of lenses. A comment would just be "prism for reading and parsing UOMs", which is just a longer version of the type signature. I could add comments anyway.